src/managers/builtin/venvManager.ts

```diff
@@ -486,39 +492,39 @@ export class VenvManager implements EnvironmentManager {
                         this.baseManager,
                     );
                     if (resolved) {
-                        // If resolved add it to the collection
-                        this.fsPathToEnv.set(p, resolved);
+                        // If resolved; add it to the venvManager collection
                         this.addEnvironment(resolved, false);
-                        if (pw && previous?.envId.id !== resolved.envId.id) {
-                            events.push(() =>
-                                this._onDidChangeEnvironment.fire({ uri: pw.uri, old: undefined, new: resolved }),
-                            );
-                        }
+                        foundEnv = resolved;
                     } else {
                         this.log.error(`Failed to resolve python environment: ${env}`);
+                        return;
                     }
                 }
+                // Given found env, add it to the map and fire the event if needed.
+                this.fsPathToEnv.set(p, foundEnv);
+                if (pw && previousEnv?.envId.id !== foundEnv.envId.id) {
+                    events.push(() =>
+                        this._onDidChangeEnvironment.fire({ uri: pw.uri, old: undefined, new: foundEnv }),
+                    );
+                }
             } else {
-                // There is NO selected venv, then try and choose the venv that is in the workspace.
-                if (sorted.length === 1) {
-                    this.fsPathToEnv.set(p, sorted[0]);
-                } else {
-                    // These are sorted by version and by path length. The assumption is that the user would want to pick
-                    // latest version and the one that is closest to the workspace.
-                    const found = sorted.find((e) => {
-                        const t = this.api.getPythonProject(e.environmentPath)?.uri.fsPath;
-                        return t && path.normalize(t) === p;
-                    });
-                    if (found) {
-                        this.fsPathToEnv.set(p, found);
-                    }
+                // Search through all known environments (e) and check if any are associated with the current project path. If so, add that environment and path in the map.
```

Comment:
@karthiknadig do we need to trigger the `this._onDidChangeEnv` for a found enviroment this way